Updates (#269)
TBD54566975/tbdex: specs/protocol/README.md

## Resource Kinds

### `Offering`

> An `Offering` is used by the PFI to describe a currency pair they have to _offer_ including the requirements, conditions, and constraints needed fulfill that offer.

Reviewer: ```suggestion
An `Offering` is used by the PFI to describe a currency pair they have to _offer_ including the requirements, conditions, and constraints needed to fulfill the offer.
```

Unclear to me the difference between `requirements` and the other two, `conditions, and constraints...`

Entirely new proposal:

```suggestion
An `Offering` is a resource created by a PFI to define requirements for a given currency pair offered for an exchange.
```

Author: i like it!

| ------------------------- | -------------------------------------------------------------------------------------------------------- | -------- | --------------------------------------------------------------------------- |
| `description`             | string                                                                                                   | Y        | Brief description of what is being offered.                                 |
| `payoutUnitsPerPayinUnit` | [`DecimalString`](#decimalstring)                                                                        | Y        | Number of payout units alice would get for 1 payin unit                     |
| `payin`                   | [`PayinDetails`](#payindetails)                                                                          | Y        | Details and options associated to the _payin_ currency                      |
| `payout`                  | [`PayoutDetails`](#payoutdetails)                                                                        | Y        | Details and options associated to the _payout_ currency                     |

> | `requiredClaims`          | [`PresentationDefinitionV2`](https://identity.foundation/presentation-exchange/#presentation-definition) | N        | Articulates the claim(s) required when submitting an RFQ for this offering. |

Reviewer: ```suggestion
| `requiredClaims`          | [`PresentationDefinitionV2`](https://identity.foundation/presentation-exchange/#presentation-definition) | N        | Claim(s) required when submitting an RFQ for this offering. |
```

| -------------- | --------------------------------- | -------- | ------------------------------------------------------ |
| `currencyCode` | string                            | Y        | ISO 3166 currency code string                          |
| `min`          | [`DecimalString`](#decimalstring) | N        | Minimum amount of currency that the offer is valid for |
| `max`          | [`DecimalString`](#decimalstring) | N        | Maximum amount of currency that the offer is valid for |
| `methods`      | [`PayoutMethod[]`](#payoutmethod) | Y        | A list of payment methods to select from               |

#### `PayinMethod`
| field                    | data type                               | required | description                                                                                                                                         |
| ------------------------ | --------------------------------------- | -------- | --------------------------------------------------------------------------------------------------------------------------------------------------- |

> | `kind`                   | string                                  | Y        | Type of payment method (i.e. `DEBIT_CARD`, `BITCOIN_ADDRESS`, `SQUARE_PAY`). considered to be a unique identifier                                   |

Reviewer 2: ```suggestion
| `kind`                   | string                                  | Y        | Unique string identifying a single kind of payment method i.e. (i.e. `DEBIT_CARD`, `BITCOIN_ADDRESS`, `SQUARE_PAY`)                                   |
```

| `group`                  | string                                  | N        | value that can be used to group specific payment methods together e.g. Mobile Money vs. Direct Bank Deposit                                         |
| `requiredPaymentDetails` | [JSON Schema](https://json-schema.org/) | N        | A JSON Schema containing the fields that need to be collected in the RFQ's selected payment methods in order to use this payment method.            |
| `fee`                    | [`DecimalString`](#decimalstring)       | N        | Fee charged to use this payment method. absence of this field implies that there is no _additional_ fee associated to the respective payment method |
| `min`                    | [`DecimalString`](#decimalstring)       | N        | minimum amount required to use this payment method.                                                                                                 |
| `max`                    | [`DecimalString`](#decimalstring)       | N        | maximum amount allowed when using this payment method.                                                                                              |

> [!IMPORTANT]

> > `kind` should be considered as a unique identifier that is used to _identify_ / _specify_ an individual payment method

Reviewer: ```suggestion
> `kind` should be a unique identifier of an individual payment method
```

Unclear to me what this would mean as an outsider coming in, I may think to myself, "_does this mean I should great a UUID?_" Perhaps we don't need this?

Reviewer 2: +1 feels duplicative to the description in the table above

> [!IMPORTANT]
> the presence of `min` or `max` in a payment method takes precedence over `min` or `max` defined at the `PaymentDetails` level.

> [!IMPORTANT]
> If `requiredPaymentDetails` is omitted, then any RFQs submitted for the respective offering must also omit `paymentDetails`. 

#### `PayoutMethod`
| field                     | data type                               | required | description                                                                                                                                         |
| ------------------------- | --------------------------------------- | -------- | --------------------------------------------------------------------------------------------------------------------------------------------------- |

> | `kind`                    | string                                  | Y        | Type of payment method (i.e. `DEBIT_CARD`, `BITCOIN_ADDRESS`, `SQUARE_PAY`). considered to be a unique identifier                                   |

Reviewer 2: ```suggestion
| `kind`                    | string                                  | Y        | Unique string identifying a single kind of payment method i.e. (i.e. `DEBIT_CARD`, `BITCOIN_ADDRESS`, `SQUARE_PAY`)                                   |
```

Forgot to include this in last pass!

> | `estimatedSettlementTime` | uint                                    | Y        | estimated time taken to settle an order. expressed in seconds                                                                                       |

Reviewer 3: Super nit: why not millis instead of seconds? I agree that the vast majority of payments with TBDex will be slower than 1 second, but I wonder if a small few may find it useful have very tight settlement times less than one second

Author: fair point @diehuxx . went with seconds because anything less than 1 second i considered to be `0` aka "instant" bc i assume no one will be using tbdex for high frequency day trading

| `group`                   | string                                  | N        | value that can be used to group specific payment methods together e.g. Mobile Money vs. Direct Bank Deposit                                         |
| `requiredPaymentDetails`  | [JSON Schema](https://json-schema.org/) | N        | A JSON Schema containing the fields that need to be collected in the RFQ's selected payment methods in order to use this payment method.            |
| `fee`                     | [`DecimalString`](#decimalstring)       | N        | Fee charged to use this payment method. absence of this field implies that there is no _additional_ fee associated to the respective payment method |
| `min`                     | [`DecimalString`](#decimalstring)       | N        | minimum amount required to use this payment method.                                                                                                 |
| `max`                     | [`DecimalString`](#decimalstring)       | N        | maximum amount allowed when using this payment method.                                                                                              |

> [!IMPORTANT]

> > `estimatedSettlementTime` is used to provide a rough estimate for the expected latency between receiving a _payin_ and settling the respective _payout_.

Reviewer: Wouldn't the estimated settlement time be a function of both the payin and the payout methods? Payin may be BTC which takes 10+ minutes when payout is debit which is instant

Reviewer 2: Could maybe clarify: is this estimate supposed to be the full settlement of the exchange (starting from `Order` message being accepted -> `Close` message being sent), or is it JUST for the payout portion (which would be a `payout initiated/payin complete` OrderStatus -> `Close` message being sent)?

And does "receiving" a payin mean the PFI has been notified that there is a payin transaction (assuming the payin method is not stored balance), or does received == payin has _settled_?

Author: > Could maybe clarify: is this estimate supposed to be the full settlement of the exchange (starting from Order message being accepted -> Close message being sent), or is it JUST for the payout portion (which would be a payout initiated/payin complete OrderStatus -> Close message being sent)?

imo it should be _just_ for the payout portion because the PFI does not control when alice pays _in_. 

> And does "receiving" a payin mean the PFI has been notified that there is a payin transaction (assuming the payin method is not stored balance), or does received == payin has settled?

received == payin has settled. in a stored balance scenario, "payin received" can be considered as instant or moot

Reviewer: It makes sense but IMO it's also relevant to the payin method. Agreed that the onus is on Alice for the payin method, but the PFI may have unique characteristics for their given payin methods. For example, consider BTC as a payin method, there is no standard rule as for how many blocks need to be mined for the payment to be considered final (even more relevant in Ethereum where reorgs are frequent), so each PFI could define their own set of rules for how long it takes for that tx to be considered final; Alice could submit a BTC tx and even see it in the blockchain but it may still not be considered settled by the PFI.

> [!IMPORTANT]
> `kind` should be considered as a unique identifier that is used to _identify_ / _specify_ an individual payment method

> [!IMPORTANT]
> the presence of `min` or `max` in a payment method takes precedence over `min` or `max` defined at the `PaymentDetails` level.

> [!IMPORTANT]
> If `requiredPaymentDetails` is omitted, then any RFQs submitted for the respective offering must also omit `paymentDetails`. 

#### `PayoutMethod`
| field                     | data type                               | required | description                                                                                                                                         |
| ------------------------- | --------------------------------------- | -------- | --------------------------------------------------------------------------------------------------------------------------------------------------- |
| `kind`                    | string                                  | Y        | Type of payment method (i.e. `DEBIT_CARD`, `BITCOIN_ADDRESS`, `SQUARE_PAY`). considered to be a unique identifier                                   |

> | `estimatedSettlementTime` | uint                                    | Y        | estimated time taken to settle an order. expressed in seconds                                                                                       |

Reviewer 4: an Alice might want to fully understand what this field means (when her wallet renders this value), that:
- this estimate is for once a tbdex Order is submitted / payin has been received | settled?
- this estimate excludes all the other steps prior to tbdex exchange, such as establishing customer relationship by kyc onboarding

Reviewer 4: ah. i guess this is mostly explained in [these lines](https://github.com/TBD54566975/tbdex/pull/269/commits/0614cf575b465581137908312c6e37c31864391c#diff-766e6d8ea5c6a7fcd675d0b65a33e80bc4aa6be519e9c2ff968ea952abf59947R173-R174) below
